src/building/bootstrapping.md

```diff
@@ -108,50 +110,92 @@ contribution [here][bootstrap-build].
-`x.py` you will see output such as:
-
-```txt
-Building stage0 std artifacts
-Copying stage0 std from stage0
-Building stage0 compiler artifacts
-Copying stage0 rustc from stage0
-Building LLVM for x86_64-apple-darwin
-Building stage0 codegen artifacts
-Assembling stage1 compiler
-Building stage1 std artifacts
-Copying stage1 std from stage1
-Building stage1 compiler artifacts
-Copying stage1 rustc from stage1
-Building stage1 codegen artifacts
-Assembling stage2 compiler
-Uplifting stage1 std
-Copying stage2 std from stage1
-Generating unstable book md files
-Building stage0 tool unstable-book-gen
-Building stage0 tool rustbook
-Documenting standalone
-Building rustdoc for stage2
-Documenting book redirect pages
-Documenting stage2 std
-Building rustdoc for stage1
-Documenting stage2 whitelisted compiler
-Documenting stage2 compiler
-Documenting stage2 rustdoc
-Documenting error index
-Uplifting stage1 rustc
-Copying stage2 rustc from stage1
-Building stage2 tool error_index_generator
-```
+### Overview
+
+This is a detailed look into the separate bootstrap stages.
+
+The convention `x.py` uses is that:
+- A `--stage N` flag means to run the stage N compiler (`stageN/rustc`).
+- A "stage N artifact" is an artifact that is _produced_ by the stage N compiler.
+- The "stage (N+1) compiler" is assembled from "stage N artifacts". This
+  process is called _uplifting_.
+
+For example, `x.py build --stage 0` means to build with the beta `rustc`, and
+`test --stage 0` isn't really meaningful (it's not running tests on your changes,
+but on `beta`, so it always fails). Similarly, `doc --stage 0` means to
+document using the beta `rustdoc`.
+
+Note that this implies the stage N compiler is _not_ the same as the compiler
+built by `build --stage N compiler/rustc` -- that's 'stage N artifacts'
+('the compiler built by stage N').
+
+In short, _stage 0 uses the stage0 compiler to create stage0 artifacts which
+will later be uplifted to stage1_.
+
+In each stage, two major steps are performed:
+
+1. `std` is compiled by the stage N compiler.
+2. That `std` is linked to programs built by the stage N compiler, including
+   the stage (N+1) compiler.
+
+This is somewhat intuitive if one thinks of the stage (N+1) compiler as "just"
+another program we are building with the stage N compiler:
+`build --stage N compiler/rustc` is linking the stage (N+1) compiler to the `std`
+built by the stage N compiler.
 
-A deeper look into `x.py`'s phases can be seen here:
+Here is a chart of a full build using `x.py`:
 
 <img alt="A diagram of the rustc compilation phases" src="../img/rustc_stages.svg" class="center" />
 
 Keep in mind this diagram is a simplification, i.e. `rustdoc` can be built at
 different stages, the process is a bit different when passing flags such as
 `--keep-stage`, or if there are non-host targets.
 
+The stage 1 artifacts are what is shipped to end-users, including `stage2/bin/rustc`.
```

Comment:
That is confusing since two stages appear here, wouldn't it be better to refer to `build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/rustc-main`?

Reply:
Hmm, the number of directories makes that a little difficult to read IMO (even if it were `build/HOST/stage1-rust/TARGET/release/rustc-main`). Maybe instead something like this?
```suggestion
The stage 1 artifacts are what is shipped to end-users. These are the same artifacts that will be uplifted to stage 2, so we sometimes say that rustup distributes stage 2.
```

Reply:
Or maybe it should just say 'distributes the stage 2 compiler'.

Reply:
I changed this to 'ships the stage 2 compiler'.